mocker.go

```diff
@@ -0,0 +1,108 @@
+package testing
+
+import (
+	"reflect"
+	"sync"
+
+	"github.com/juju/loggo"
+)
+
+// NewCallMocker returns a CallMocker which will log calls and results
+// utilizing the given logger.
+func NewCallMocker(logger loggo.Logger) *CallMocker {
+	return &CallMocker{
+		logger:  logger,
+		results: make(map[string][]*callMockReturner),
+	}
+}
+
+// CallMocker is a tool which allows tests to dynamically specify
+// results for a given set of input parameters.
+type CallMocker struct {
+	Stub
+
+	logger  loggo.Logger
+	results map[string][]*callMockReturner
+}
+
+// MethodCall logs the call to a method and any results that will be
+// returned. It returns the results previously specified by the Call
+// function. If no results were specified, the returned slice will be
+// nil.
+func (m *CallMocker) MethodCall(receiver interface{}, fnName string, args ...interface{}) []interface{} {
+	m.Stub.MethodCall(receiver, fnName, args...)
+	m.logger.Debugf("Call: %s(%v)", fnName, args)
+	results := m.Results(fnName, args...)
+	m.logger.Debugf("Results: %v", results)
+	return results
+}
+
+// Results returns any results previously specified by calls to the
+// Call method. If there are no results, the returned slice will be
+// nil.
+func (m *CallMocker) Results(fnName string, args ...interface{}) []interface{} {
+	for _, r := range m.results[fnName] {
+		if reflect.DeepEqual(r.args, args) == false {
+			continue
+		}
+		r.logCall()
+		return r.retVals
+	}
+	return nil
```

Comment:
I wonder if this should be an error, that is, shouldn't the test author be required to specify each possible call up front? Otherwise tests could end up passing unintentionally.